docs/frameworks/nestjs.md

```diff
@@ -0,0 +1,161 @@
+
+## Hello World application
+
+Before going into Featurevisor integration, let's create a simple Hello World [Nest.js](https://nestjs.com/) application.
+
+We start by installing the Nest CLI:
+
+```bash
+$ npm install -g @nestjs/cli
+```
+
+Then we create a new Nest.js project:
+
+```bash
+$ nest new featurevisor-nest-app
+$ cd featurevisor-nest-app
+```
+
+## Featurevisor integration
+
+We install the Featurevisor SDK first:
+
+```bash
+$ npm install --save @featurevisor/sdk
+```
+
+We can now create an instance of the SDK and use it in our application:
+
+```typescript
+// main.ts
+import { NestFactory } from '@nestjs/core';
+import { AppModule } from './app.module';
+import { createInstance, FeaturevisorInstance } from '@featurevisor/sdk';
+
+async function bootstrap() {
+  const app = await NestFactory.create(AppModule);
+
+  const DATAFILE_URL = 'https://cdn.yoursite.com/datafile.json';
+  const REFRESH_INTERVAL = 60 * 5; // every 5 minutes
+
+  const f: FeaturevisorInstance = createInstance({
+    datafileUrl: DATAFILE_URL,
+    refreshInterval: REFRESH_INTERVAL,
+  });
+
+  app.use((req, res, next) => {
+    req.f = f;
+
+    next();
+  });
+
+  await app.listen(3000);
+}
+bootstrap();
+```
+
+## middleware
+
+To use the same Featurevisor SDK instance in all of our routes, we can create a custom middleware in Nest.js.
+
+Create a new file featurevisor.middleware.ts:
+
+```typescript
+// featurevisor.middleware.ts
+import { Injectable, NestMiddleware } from '@nestjs/common';
+import { Request, Response, NextFunction } from 'express';
+import { FeaturevisorInstance, createInstance } from '@featurevisor/sdk';
+
+@Injectable()
+export class FeaturevisorMiddleware implements NestMiddleware {
+  private readonly f: FeaturevisorInstance;
+
+  constructor() {
+    const DATAFILE_URL = 'https://cdn.yoursite.com/datafile.json';
+    const REFRESH_INTERVAL = 60 * 5; // every 5 minutes
+
+    this.f = createInstance({
+      datafileUrl: DATAFILE_URL,
+      refreshInterval: REFRESH_INTERVAL,
+    });
+  }
+
+  use(req: Request, res: Response, next: NextFunction) {
+    req.f = this.f;
+
+    next();
+  }
+}
+```
+
+And then update the main.ts file to use the middleware:
+
+```typescript
+Copy code
+// main.ts
+import { NestFactory } from '@nestjs/core';
+import { AppModule } from './app.module';
+import { FeaturevisorMiddleware } from './featurevisor.middleware';
+
+async function bootstrap() {
+  const app = await NestFactory.create(AppModule);
+
+  app.use(new FeaturevisorMiddleware().use);
+
+  await app.listen(3000);
+}
+bootstrap();
+```
+
+## Using Featurevisor in a route
+
+Now you can use the Featurevisor SDK instance in your routes:
+
+```typescript
+// cats.controller.ts
+import { Controller, Get, Req } from '@nestjs/common';
+import { Request } from 'express';
+
+@Controller('cats')
+export class CatsController {
+  @Get()
+  getCats(@Req() req: Request) {
+    const { f } = req;
```

Comment:
NestJS also has a Providers concept: https://docs.nestjs.com/providers

Do you think we will be better of using that instead of injected FeaturevisorInstance to every Request object?

Reply:
I agree with @fahad19, modifying requests through a middleware makes the code less transparent, and given that none of the values provided by the `Request` nor the `Resolve` objects are required for the `FeaturevisorInstance`, it makes even less sense to put it there. It also creates problems for whoever uses other web frameworks rather than Express as the underlying HTTP framework, as they would have to do declaration merging differently from what is available in this doc.


Registering the library as a `Provider` would make more sense, as it allows devs to inject it exclusively into the controllers where they make use of the library.

Moreover, it is [recommended](https://docs.nestjs.com/fundamentals/custom-providers#value-providers-usevalue) by the NestJS docs to register a provider using `useValue` when your purpose is to put external libraries into the container.

This alternative solution would look like this

```ts
import { FeaturevisorInstance, createInstance as createFeaturevisor } from '@featurevisor/sdk';

const FeaturevisorProvider = { 
  provide: 'Featurevisor', 
  useValue: createFeaturevisor({
    datafileUrl: 'https://cdn.yoursite.com/datafile.json',
    refreshInterval: 60 * 5,  // every 5 minutes
  }) 
};

@Module({
  imports: [CoolModule],
  providers: [FeaturevisorProvider],
})
export class AppModule {}
```

Another benefit of this approach when compared to using middlewares would be not having to maintain the forwarding of the options to the instance creator.

Reply:
I believe we should also provide guidance for the usage of dynamic values while creating the instance. A clear use case for that would be when devs get the `datafileUrl` value out of a config service.

Such case could be handled by registering the provider using `useFactory`, which allows for creating providers dynamically.

```tsx
import { FeaturevisorInstance, createInstance as createFeaturevisor } from '@featurevisor/sdk';

const FeaturevisorProvider = { 
  provide: 'Featurevisor', 
  useFactory: async (configProvider: OptionsProvider) => {
    const featureOptions = await configProvider.getFeatureOptions();
    return createFeaturevisor(featureOptions);
  },
  inject: [ConfigProvider],
};

@Module({
  imports: [CoolModule],
  providers: [FeaturevisorProvider],
})
export class AppModule {}
```